### PR TITLE
[Wave Collect][Xero] Import tracking categories

### PR DESCRIPTION
**Pull request by mananjadhav**
### Fixed Issues

$ https://github.com/Expensify/App/issues/39732
PROPOSAL:

### Tests

**Pre-requisite:**
1. Xero account setup in an existing workspace.
2. You need to have 2 tracking categories named "Region" and "Cost centers",  to set their mapping.
3. On your Xero account, select the Demo company organization.
4. Go to Accounting > Advanced, select "Tracking categories".
5. You should see something named "Region", in reality it can be any name that's why you must show/hide the option depending on the name of the tracking category.
6. Add a new tracking category named "Cost centers".

https://github.com/Expensify/App/assets/3069065/48094376-10db-4e18-b4fa-586d0a4ee134

**Steps**
1. Login with a user who already has the Xero account setup.
2. Go to Workspace having Xero connection -> Accounting -> Import.
3. Click on `Tracking Categories`
4. Verify that you're able to see `Import` toggle switch.
5. If the import toggle is off, then you shouldn't see any other option in the page.
6. Switch the `Import` to On, in the data you can verify that `policy_<policyID>.connections.xero.config.importTrackingCategories` is set to true. On switching off, it is set to false.
8. Once turned on you should see two items `Map Xero cost centers to` and `Map Xero regions to`.
9. Click on each of the menu items and perform the following steps:
       a. You should see the title of the page as the `Map Xero <fieldname> to`.
       b. You should see two options `Xero contact default` and `Tags` as two menu items.
       c. On selecting any one of the options, you should be redirected back to the tracking categories page, and the selected option should show up against each field title.
       d. In `policy_<policyID>.connections.xero.mappings.trackingCategory_<trackingCategoryId>` you should see the selected value , ie, `DEFAULT` for `Xero contact default` and `TAG` for `Tags`.
       e. The trackingCategoryId should be based on the `id` field of  `policy_<policyID>.connections.xero.trackingCategories.name` matched to cost centers and regions respectively.

For all the data verification related info, go to Chrom Dev Tools -> Application -> Storage -> OnyxDB -> Search for your policy_<policyID>. Check out the data in the connections object as shown in the screenshot below.

image

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as Tests.

### QA Steps

Same as Tests.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/3069065/bc492e12-81d7-4662-ab37-602b690fd9d5

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/3069065/a34a23fd-fd59-48cb-8fb3-28e257b015bf

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/3069065/104b5d0b-bda5-449e-adaa-57f088a07e44

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/3069065/51a1aa66-97e3-4803-ba41-620924d35b6d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/3069065/ea3dd2f9-3147-4921-b0ff-d748c80bb12e

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/3069065/28197ed0-3d81-40d0-ace1-12e36ba1caac

</details>

### CODE REVIEW

**Comment by lakchote** [comment]
Could we have the Spanish translations for these, please?

```
trackingCategories: 'Tracking categories',
trackingCategoriesDescription: 'Choose whether to import tracking categories and see where they are displayed.',
mapXeroCostCentersTo: 'Map Xero cost centers to',
mapXeroRegionsTo: 'Map Xero regions to',
mapXeroCostCentersToDescription: 'Choose where to map cost centers to when exporting to Xero.',
mapXeroRegionsToDescription: 'Choose where to map employee regions when exporting expense reports to Xero.',
xeroContactDefault: 'Xero contact default',
```

File: src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx
Comment (rushatgabhane) [inline src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx:57]: Please use `ConnectionLayout` component

**Review by rushatgabhane — CHANGES_REQUESTED** [review]

File: src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx
Comment (mananjadhav) [reply src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx]: Updated.

**Comment by mananjadhav** [comment]
On Hold for https://github.com/Expensify/App/pull/41561. Waiting for `ConnectionLayout` padding fix to get merged so that I can create the screencasts for the mapping screens.

**Comment by mananjadhav** [comment]
@Expensify/design For your review.

File: src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx
Comment (lakchote) [inline src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx:41]: Why do you have different logic for `Cost centers` and `Region`? 

With `if (costCenterCategoryValue) {` and `trackingCategories?.find((category)` ?

**Comment by rushatgabhane** [comment]
@mananjadhav i can't see cost center and region

image

File: src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx
Comment (mananjadhav) [reply src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx]: Forgot to refactor this one line. Updated.

**Comment by mananjadhav** [comment]
@rushatgabhane Did you check the pre-requisite in the Test steps? and did you try to select the `Demo company` ?

> Pre-requisite:
> 
> 1. Xero account setup in an existing workspace.
> 2. You need to have 2 tracking categories named "Region" and "Cost centers", to set their mapping.
> 3. On your Xero account, select the Demo company organization.
> 4. Go to Accounting > Advanced, select "Tracking categories".
> 5. You should see something named "Region", in reality it can be any name that's why you must show/hide the option depending on the name of the tracking category.
> 6. Add a new tracking category named "Cost centers".

**Comment by lakchote** [comment]
> @mananjadhav i can't see cost center and region
> 

To follow up with @mananjadhav, please see [this](https://expensify.slack.com/archives/C036QM0SLJK/p1714625508724419?thread_ts=1713799302.291019&cid=C036QM0SLJK) discussion for context on how to enable them.

**Comment by mananjadhav** [comment]
@lakchote I didn't understand what you meant? The link also doesn't take me to any specific thread. I already added the pre-requisite on how to add them based on your instructions in slack. Is there anything else that's expected from me?

**Comment by shawnborton** [comment]
Design-wise this looks good to me.

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
Bug:  Category is not mapped properly

1. In Old dot, go to the connection > Coding 
2.  Map Xero regions to - "Report Field (header level)"

This is because the category options are hardcoded to `Tags` and `Default`. But I selected `Report Field`


image


Screenshot 2024-05-06 at 17 28 55

File: src/CONST.ts
Comment (rushatgabhane) [inline src/CONST.ts:1293]: These shouldn't be hardcoded, right?

Comment (mananjadhav) [reply src/CONST.ts:1293]: @lakchote I had asked if these will be hardcoded. Thread [here](https://expensify.slack.com/archives/C036QM0SLJK/p1714638671089049?thread_ts=1713799302.291019&cid=C036QM0SLJK).

**Comment by mananjadhav** [comment]
> This is because the category options are hardcoded to Tags and Default. But I selected Report Field

Got it. We'll be adding this in future.  But at the moment I think it should show blank here then? @trjExpensify @lakchote ?

**Comment by rushatgabhane** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by rushatgabhane — APPROVED** [review]
@lakchote LGTM if you agree with https://github.com/Expensify/App/pull/41441#issuecomment-2096399216

File: src/CONST.ts
Comment (lakchote) [reply src/CONST.ts:1293]: These will be hardcoded yes.

**Comment by lakchote** [comment]
> > This is because the category options are hardcoded to Tags and Default. But I selected Report Field
> 
> Got it. We'll be adding this in future. But at the moment I think it should show blank here then? @trjExpensify @lakchote ?

Report field will be added once we'll port `Control` workspaces to NewDot. It's a Control feature.

I'd move in favour of showing blank there in the meantime.

**Comment by mananjadhav** [comment]
Thanks for the comment @lakchote. I'll push an update in a while for the Report field.

**Comment by mananjadhav** [comment]
@rushatgabhane @lakchote Added a fix to show blank for the report field. I didn't refactor too much because we are anyway going remove it later.

Screenshot 2024-05-07 at 9 47 39 PM

**Review by lakchote — APPROVED** [review]

**Comment by lakchote** [comment]
I have merged by mistake (since we're on merge freeze). Since it's behind beta, and it should not influence any user behaviour we don't need to revert it.

I've created a new `xero-merge-freeze` branch for the next Xero PRs, people will need to use it as a base instead of `main` in the meantime to not block progress.

**Comment by lanitochka17** [comment]
Tell us about PR Pre-props:
step 4 cannot be completed because we do not see "Tracking categories"
Go to Accounting > Advanced, select Tracking Categories.
How do we set up an account to complete these steps?
thank you in advance


https://github.com/Expensify/App/assets/78819774/4d56ce08-77d3-4a7d-b3fb-1de2ec19d8b4

**Comment by mananjadhav** [comment]
@lanitochka17 Did you try the steps mentioned in the pre-requisites? Please find attached video on navigating through Xero. You should see Demo Company when you create a new account on Xero, and use that for syncing when you test it.

https://github.com/Expensify/App/assets/3069065/48094376-10db-4e18-b4fa-586d0a4ee134

**Comment by m-natarajan** [comment]
@mananjadhav Thanks for sharing the video again. I have a question, How to verify the PR step 6 and 8.d and e? Is this can only be done internally?

**Comment by mananjadhav** [comment]
You should be able to view this in the application data onyxdb in Chrom Dev tools. I've added a note in the steps along with the screenshot.

File: src/pages/workspace/accounting/xero/XeroMapCostCentersToConfigurationPage.tsx
Comment (ahmedGaber93) [inline src/pages/workspace/accounting/xero/XeroMapCostCentersToConfigurationPage.tsx:35]: Coming from https://github.com/Expensify/App/issues/64288, we need to `enablePolicyReportFields` if user set/update any of tracking options to `Report Fields`

**Pull request merged**
